### PR TITLE
Autodata glue for nunit3 (issue #522)

### PR DESCRIPTION
**Pull request by hackle**
This is ported over from https://github.com/hackle/AutoFixture.NUnit3.AutoDataConnector following issue: https://github.com/AutoFixture/AutoFixture/issues/522

In short it introduces AutoDataAttribute which supports FrozenAttribute, with an effort at making everything as testable as possible while keeping code ported over from NUnit (in particular TestCaseAttribute).

Points / concerns:

1. No nuget spec yet - NUnit3 has different packages for different versions of .NET, which may require corresponding design, but that can complicate this first pull request. The current project (and therefore resultant dll) only targets .NET4.0

2. No changes to the build script for the same reason

Thanks.

### CODE REVIEW

File: BuildRelease.msbuild [hunk 1]
```diff
@@ -107,7 +107,7 @@
         <ItemGroup>
           <XunitTestAssemblies
               Include="Src\*Test\bin\$(Configuration)\*Test.dll"
-              Exclude="Src\AutoFixture.NUnit2.*Test\bin\$(Configuration)\*.dll" />
+              Exclude="Src\AutoFixture.NUnit*.*Test\bin\$(Configuration)\*.dll" />
```
Comment (moodmosaic) [inline BuildRelease.msbuild]: Is this change necessary?

File: Src/AutoFixture.NUnit3/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("AutoFixture.NUnit3")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("Ploeh")]
+[assembly: AssemblyProduct("AutoFixture.NUnit3")]
+[assembly: AssemblyCopyright("Copyright © Ploeh 2011")]
```
Comment (moodmosaic) [inline Src/AutoFixture.NUnit3/Properties/AssemblyInfo.cs]: Is this line accurate enough? :confused:

File: Src/AutoFixture.NUnit3/ITypedValueProvider.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+namespace Ploeh.AutoFixture.NUnit3
+{
+    public interface ITypedValueProvider
```
Comment (moodmosaic) [inline Src/AutoFixture.NUnit3/ITypedValueProvider.cs]: What's the purpose of `ITypedValueProvider`?

**Comment by moodmosaic** [comment]
Thank you for your contribution!

AFAICT, there are public types introduced in this Glue Library that don't exist in AutoFixture.NUnit2. Do we really need to support and maintain them? :confused: 

Unless @ploeh, @adamchester, or @ecampidoglio have a different opinion, can this Pull Request be broken into separate smaller ones? ([Similar to what we did with AutoFixture.Xunit2](https://github.com/AutoFixture/AutoFixture/pull/294#issuecomment-82385218).)

File: BuildRelease.msbuild (hunk 1 above)
Comment (hackle) [inline BuildRelease.msbuild]: This is to exclude AutoFixture.NUnit3 unit tests from breaking the build.

File: Src/AutoFixture.NUnit3/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (hackle) [inline Src/AutoFixture.NUnit3/Properties/AssemblyInfo.cs]: Not enough I guess I'll update it. Missed it when copying from AutoFixture.NUnit2...

File: Src/AutoFixture.NUnit3/ITypedValueProvider.cs (hunk 3 above)
Comment (hackle) [inline Src/AutoFixture.NUnit3/ITypedValueProvider.cs]: Hmm so this might be a bad name. The purpose of it is to expose capacities to provide values with different purposes - hence the methods CreateValue(Type) and CreateFrozenValue(Type).

I agree on the concern of exposure of public interfaces which may cause confusion.

**Comment by hackle** [comment]
Thanks for your timely review @moodmosaic I just skimped through the AutoFixture.Xunit2 thread and think that makes sense. (@ploeh actually referenced it in issue #522 so my bad to have not followed up) Is it as easy as following the same path and starting with small requests?

**Comment by moodmosaic** [comment]
@hackle :+1:

>Is it as easy as following the same path and starting with small requests?

This way it was easier with AutoFixture.Xunit2, so it should also be easier with AutoFixture.NUnit3. In case it helps, #377 and #378 were the first merged Pull Requests towards AutoFixture.Xunit2. :wink:

**Comment by ploeh** [comment]
Thank you for your interest in contributing to AutoFixture :+1: 

As @moodmosaic has already suggested, I'd appreciate if this pull request could be split into several smaller pull requests. If not, I'm afraid the review may miss important points, and instead devolve into [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality). The addition of support for NUnit 3 is, I think, too important for that.

Could the first step be to add a new, empty Visual Studio solution?

For the reasons outlined above, I'm going to close this pull request. It's important to me, however, to underscore that I welcome the contribution, and that I think it can become a valuable addition to AutoFixture. I'm closing this pull request for the sole reason that it's too big. Apart from that, you should read no judgement into this action, as I haven't otherwise looked at the code yet.

Please don't be discouraged that this pull request is closed. We still eagerly await your contribution to this new feature; we only need them in smaller instalments.
